File: python_app/agents/weather.py
import requests

OPEN_METEO_URL = "https://api.open-meteo.com/v1/forecast"

def _round(n):
    return round(n * 10) / 10

def get_weather(lat, lon):
    params = {
        "latitude": lat,
        "longitude": lon,
        "current": "temperature_2m",
        "hourly": "precipitation_probability",
        "timezone": "auto",
    }
    try:
        r = requests.get(OPEN_METEO_URL, params=params, timeout=20)
    except Exception:
        return {"temperature": None, "precipProb": None}
    if r.status_code != 200:
        return {"temperature": None, "precipProb": None}
    ct = r.headers.get("content-type", "").lower()
    if "application/json" not in ct:
        return {"temperature": None, "precipProb": None}
    try:
        j = r.json()
    except Exception:
        return {"temperature": None, "precipProb": None}
    t = j.get("current", {}).get("temperature_2m")
    ct_time = j.get("current", {}).get("time")
    times = j.get("hourly", {}).get("time", [])
    probs = j.get("hourly", {}).get("precipitation_probability", [])
    p = None
    if ct_time and times and probs:
        key = ct_time[:13] + ":00" if isinstance(ct_time, str) and len(ct_time) >= 13 else None
        idx = times.index(key) if key in times else (times.index(ct_time) if ct_time in times else -1)
        if idx >= 0 and isinstance(probs[idx], (int, float)):
            p = int(round(probs[idx]))
    return {"temperature": _round(t) if isinstance(t, (int, float)) else None, "precipProb": p if isinstance(p, int) else None}

File: src/agents/weather.js
const OPEN_METEO_URL = "https://api.open-meteo.com/v1/forecast";

function round(n) {
  return Math.round(n * 10) / 10;
}

export async function getWeather(lat, lon) {
  const u = new URL(OPEN_METEO_URL);
  u.searchParams.set("latitude", String(lat));
  u.searchParams.set("longitude", String(lon));
  u.searchParams.set("current", "temperature_2m");
  u.searchParams.set("hourly", "precipitation_probability");
  u.searchParams.set("timezone", "auto");
  let res;
  try {
    res = await fetch(u.toString(), { headers: { "Accept": "application/json" } });
  } catch {
    return { temperature: null, precipProb: null };
  }
  if (!res.ok) return { temperature: null, precipProb: null };
  const ctype = (res.headers.get("content-type") || "").toLowerCase();
  if (!ctype.includes("application/json")) return { temperature: null, precipProb: null };
  let j;
  try {
    j = await res.json();
  } catch {
    return { temperature: null, precipProb: null };
  }
  const t = j?.current?.temperature_2m;
  const ct = j?.current?.time;
  const times = j?.hourly?.time || [];
  const probs = j?.hourly?.precipitation_probability || [];
  let p = null;
  if (ct && times.length && probs.length) {
    let key = null;
    if (typeof ct === "string" && ct.length >= 13) key = ct.slice(0, 13) + ":00";
    let idx = key ? times.indexOf(key) : -1;
    if (idx < 0) idx = times.indexOf(ct);
    if (idx >= 0 && typeof probs[idx] === "number") p = probs[idx];
  }
  return { temperature: typeof t === "number" ? round(t) : null, precipProb: typeof p === "number" ? Math.round(p) : null };
}